crates/nu-cli/src/completions/completer.rs

```diff
@@ -140,305 +182,278 @@ impl NuCompleter {
 
         let config = self.engine_state.get_config();
 
-        let outermost_block = parse(&mut working_set, Some("completer"), line.as_bytes(), false);
+        let block = parse(&mut working_set, Some("completer"), line.as_bytes(), false);
+        let Some(element_expression) = block.find_map(&working_set, &|expr: &Expression| {
+            find_pipeline_element_by_position(expr, &working_set, pos)
+        }) else {
+            return vec![];
+        };
 
-        // Try to get the innermost block parsed (by span) so that we consider the correct context/scope.
-        let target_block = working_set
-            .delta
-            .blocks
-            .iter()
-            .filter_map(|block| match block.span {
-                Some(span) if span.contains(pos) => Some((block, span)),
-                _ => None,
-            })
-            .reduce(|prev, cur| {
-                // |(block, span), (block, span)|
-                match cur.1.start.cmp(&prev.1.start) {
-                    core::cmp::Ordering::Greater => cur,
-                    core::cmp::Ordering::Equal if cur.1.end < prev.1.end => cur,
-                    _ => prev,
+        let flattened = flatten_expression(&working_set, element_expression);
+        let mut spans: Vec<String> = vec![];
+
+        for (flat_idx, (span, shape)) in flattened.iter().enumerate() {
+            let is_passthrough_command = spans
+                .first()
+                .filter(|content| content.as_str() == "sudo" || content.as_str() == "doas")
+                .is_some();
```

Comment:
One obstacle before fully pattern matching based completion is this one, which I find not particularly useful.

Reply:
It seems useful to me, although limiting it to sudo and doas isn't great. I guess you could type out the command without `sudo` to get completions, then go back and add `sudo` at the beginning, but it's more convenient to not have to do that.

Reply:
Yeah, I get your point. Could use some special checks besides pattern matching, since all it requires is just commands, pretty common stuff, maybe not a big deal.

Another question, does it make sense to run an internal command with sudo?

Reply:
It has a special branch for stuff like `sudo use` `sudo ls`, I'm not sure why.
https://github.com/nushell/nushell/blob/1349187e17b20e554bcc1a37ce6e10494b23c6bc/crates/nu-cli/src/completions/completer.rs#L287

Reply:
Huh. I have no idea why the `is_passthrough_command && flat_idx > 1` check is there. That whole branch is a bit fishy, honestly, because https://github.com/nushell/nushell/issues/14852 happens there.

Reply:
> Huh. I have no idea why the `is_passthrough_command && flat_idx > 1` check is there. That whole branch is a bit fishy, honestly, because https://github.com/nushell/nushell/issues/14852 happens there.

Yeah, that's one of the reasons for a pattern matching based rewriting. It would be much clearer, if sudo and doas only live with externals.

Would you like to review this PR, or can you help to ping some experts of this section?

Reply:
I think this is the PR that introduced that branch https://github.com/nushell/nushell/pull/8094

Reply:
> I think this is the PR that introduced that branch https://github.com/nushell/nushell/pull/8094

I believe it tried to follow the exact logic without the prefix.

I never want to run an internal command with sudo, is there any legitimate use case?

Reply:
```nushell
sudo rm -r foo
```

Reply:
> ```nushell
> 
> sudo rm -r foo
> 
> ```

Oh, I see, windows doesn't have an external rm. hmmm

Reply:
very recent versions of windows have sudo https://github.com/microsoft/sudo or allow you to build/download. rm is built-in command.

Reply:
Ok, let me rephrase the question,

Do we want to complete some internal flags after sudo? If the only thing we care about is the command itself, easy to do in a relative elegant way.

Reply:
I'm not sure. I don't see why we'd complete flags for sudo without custom completions.

Reply:
I think I've got a clear view of how to refactor that in future work, regardless of how it evolves to that in the first place.